src/adsbinarynode.jl

```diff
@@ -0,0 +1,402 @@
+mutable struct ADSBinaryNode{N, M, T <: Real}
+    s::ADSDomain{N, T}
+    t::T
+    x::SVector{M, TaylorN{T}}
+    p::SVector{M, Taylor1{TaylorN{T}}}
+    depth::Int
+    parent::Union{Nothing, ADSBinaryNode{N, M, T}}
+    left::Union{Nothing, ADSBinaryNode{N, M, T}}
+    right::Union{Nothing, ADSBinaryNode{N, M, T}}
+    # Inner constructor (root node by default)
+    function ADSBinaryNode{N, M, T}(
+        s::ADSDomain{N, T}, t::T, x::SVector{M, TaylorN{T}},
+        p::SVector{M, Taylor1{TaylorN{T}}}, depth::Int = 0,
+        parent = nothing, l = nothing, r = nothing
+        ) where {N, M, T <: Real}
+        new{N, M, T}(s, t, x, p, depth, parent, l, r)
+    end
+end
+
+# Outer constructor (root node)
+function ADSBinaryNode(
+    s::ADSDomain{N, T}, t::T, x::SVector{M, TaylorN{T}},
+    p::SVector{M, Taylor1{TaylorN{T}}}
+    ) where {N, M, T <: Real}
+    return ADSBinaryNode{N, M, T}(s, t, x, p)
+end
+
+# Print method for ADSBinaryNode
+# Example:
+# x: [1.0, 0.0, 0.0, 1.224744871391589]
+# s: [-1.0, 1.0]×[-1.0, 1.0]
+# t: 0.0
+function show(io::IO, n::ADSBinaryNode{N, M, T}) where {N, M, T <: Real}
+    print(io, "x: ", constant_term.(n.x), "\ns: ", n.s, "\nt: ", n.t)
+end
+
+# Left (right) child constructors
+function leftchild!(
+    parent::ADSBinaryNode{N, M, T}, s::ADSDomain{N, T}, t::T,
+    x::SVector{M, TaylorN{T}}, p::SVector{M, Taylor1{TaylorN{T}}}
+    ) where {N, M, T <: Real}
+    isnothing(parent.left) || error("Left child is already assigned")
+    node = ADSBinaryNode{N, M, T}(s, t, x, p, parent.depth + 1, parent)
+    parent.left = node
+end
+
+function rightchild!(
+    parent::ADSBinaryNode{N, M, T}, s::ADSDomain{N, T}, t::T,
+    x::SVector{M, TaylorN{T}}, p::SVector{M, Taylor1{TaylorN{T}}}
+    ) where {N, M, T <: Real}
+    isnothing(parent.right) || error("Right child is already assigned")
+    node = ADSBinaryNode{N, M, T}(s, t, x, p, parent.depth + 1, parent)
+    parent.right = node
+end
+
+# AbstractTrees interface
+function children(node::ADSBinaryNode{N, M, T}) where {N, M, T <: Real}
+    if isnothing(node.left) && isnothing(node.right)
+        ()
+    elseif isnothing(node.left) && !isnothing(node.right)
+        (node.right,)
+    elseif !isnothing(node.left) && isnothing(node.right)
+        (node.left,)
+    else
+        (node.left, node.right)
+    end
+end
+
+printnode(io::IO, n::ADSBinaryNode{N, M, T}) where {N, M, T <: Real} = print(io, "t: ", n.t)
+
+nodevalue(n::ADSBinaryNode{N, M, T}) where {N, M, T <: Real} = (n.s, n.t, n.x, n.p)
+
+ParentLinks(::Type{<:ADSBinaryNode{N, M, T}}) where {N, M, T <: Real} = StoredParents()
+
+parent(n::ADSBinaryNode{N, M, T}) where {N, M, T <: Real} = n.parent
+
+NodeType(::Type{<:ADSBinaryNode{N, M, T}}) where {N, M, T <: Real} = HasNodeType()
+nodetype(::Type{<:ADSBinaryNode{N, M, T}}) where {N, M, T <: Real} = ADSBinaryNode{N, M, T}
+
+# For TreeIterator
+IteratorEltype(::Type{<:TreeIterator{ADSBinaryNode{N, M, T}}}) where {N, M, T <: Real} = HasEltype()
```

Comment:
I agree that this may be redundant, but we should leave it as it is following this [note](https://juliacollections.github.io/AbstractTrees.jl/stable/iteration/#Interface) from `AbstractTrees`.

Reply:
I agree we should keep it; I would only suggest to add a comment in the code explaining this together with the link you posted

Reply:
Done.